### ISSUE TITLE
is CORS broken?

### ISSUE THREAD

**Issue by neoel** [dunglas/mercure#556]
Trying to setup mercure with cors, but I can't seem to get it to work. 

Think it's been like that for a while, since #442 has been closed with wontfix by a stale bot we'd have to keep opening issues? or get some understanding in the issue

Using this caddyfile with the docker image:
```
# Learn how to configure the Mercure.rocks Hub on https://mercure.rocks/docs/hub/config
{
    # Debug mode (disable it in production!)
    {$DEBUG:debug}
    # HTTP/3 support
    servers {
        protocol {
            experimental_http3
        }
    }
}

{$SERVER_NAME:localhost}

log

route {
    redir / /.well-known/mercure/ui/
    encode zstd gzip

    mercure {
        # Transport to use (default to Bolt)
        transport_url {$MERCURE_TRANSPORT_URL:bolt://mercure.db}
        # Publisher JWT key
        publisher_jwt {env.MERCURE_PUBLISHER_JWT_KEY} {env.MERCURE_PUBLISHER_JWT_ALG}
        # Subscriber JWT key
        subscriber_jwt {env.MERCURE_SUBSCRIBER_JWT_KEY} {env.MERCURE_SUBSCRIBER_JWT_ALG}
        # Permissive configuration for the development environment
        cors_origins http://votes.wolbodo/
        publish_origins *
        demo
        anonymous
        subscriptions
        # Extra directives
        {$MERCURE_EXTRA_DIRECTIVES}
    }

    respond /healthz 200

    respond "Not Found" 404
}
```

**Comment by dunglas** [comment]
Can you post the exact error you get in the browser, the URL of the frontend app and the JS code please?

Also, I created a repo to test the various CORS-related issues: https://github.com/dunglas/mercure-reproducer-cors

Could you try to open a PR on this repo to add a reproducer?

Currently, all reported problems were config problems, not bugs in Mercure.

**Comment by neoel** [comment]
I "fixed" the issue by adding the headers in the Caddyfile:

inside the route config (not really familiar with caddy)
```
    header Access-Control-Allow-Origin http://votes.wolbodo
    header Access-Control-Allow-Credentials true
```
which works.

In the browser (firefox 91) I'm getting the generic unhelpful cors error. 

![image](https://user-images.githubusercontent.com/1440704/131383567-32128d4d-1329-4bc0-9328-09a83e0ca966.png)

The frontend url is "votes.wolbodo", which I'm running locally, so it doesn't have a proper TLD

The code in the frontend is:
```js
onMount(async () => {
    const url = new URL('http://mercure.wolbodo/.well-known/mercure');
    url.searchParams.append('topic', 'http://votes.wolbodo/member');
    const eventSource = new EventSource(url, { withCredentials: true });

    eventSource.addEventListener("notice", function(e) {
      console.log(e.data)
    })

    eventSource.addEventListener("update", function(e) {
      console.log(e.data)
    })

    eventSource.addEventListener("message", function(e) {
      console.log(e.data)
    })

    return () => eventSource.close()
  })
```

**Comment by neoel** [comment]
I see different behaviour on my docker-compose file vs your test setup:

My setup:
```bash
➜  ~ curl -N -v 'http://mercure.wolbodo/.well-known/mercure?topic=http%3A%2F%2Fvotes.wolbodo%2Fmember' -H 'User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0' -H 'Accept: text/event-stream' -H 'Accept-Language: en-US,en;q=0.5' --compressed -H 'Origin: http://votes.wolbodo' -H 'Connection: keep-alive' -H 'Referer: http://votes.wolbodo/' -H 'Pragma: no-cache' -H 'Cache-Control: no-cache'
*   Trying 172.18.0.5:80...
* TCP_NODELAY set
* Connected to mercure.wolbodo (172.18.0.5) port 80 (#0)
> GET /.well-known/mercure?topic=http%3A%2F%2Fvotes.wolbodo%2Fmember HTTP/1.1
> Host: mercure.wolbodo
> Accept-Encoding: deflate, gzip, br
> User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0
> Accept: text/event-stream
> Accept-Language: en-US,en;q=0.5
> Origin: http://votes.wolbodo
> Connection: keep-alive
> Referer: http://votes.wolbodo/
> Pragma: no-cache
> Cache-Control: no-cache
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Cache-Control: private, no-cache, no-store, must-revalidate, max-age=0
< Connection: keep-alive
< Content-Security-Policy: default-src 'self' mercure.rocks cdn.jsdelivr.net
< Content-Type: text/event-stream
< Expire: 0
< Pragma: no-cache
< Server: Caddy
< X-Accel-Buffering: no
< X-Content-Type-Options: nosniff
< X-Frame-Options: DENY
< X-Xss-Protection: 1; mode=block
< Date: Mon, 30 Aug 2021 18:10:48 GMT
< Transfer-Encoding: chunked
< 
:
```

Your setup:
```bash
➜  ~ curl -N -v -k 'https://domain1.localhost/.well-known/mercure?topic=foo' -H 'User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0' -H 'Accept: text/event-stream' -H 'Accept-Language: en-US,en;q=0.5' --compressed -H 'Origin: https://domain2.localhost:4443' -H 'Connection: keep-alive' -H 'Referer: https://domain2.localhost:4443/' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: cross-site' -H 'Pragma: no-cache' -H 'Cache-Control: no-cache' -H 'TE: trailers'
*   Trying ::1:443...
* TCP_NODELAY set
* Connected to domain1.localhost (::1) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: [NONE]
*  start date: Aug 30 18:03:02 2021 GMT
*  expire date: Aug 31 06:03:02 2021 GMT
*  issuer: CN=Caddy Local Authority - ECC Intermediate
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x55a58bd36e10)
> GET /.well-known/mercure?topic=foo HTTP/2
> Host: domain1.localhost
> accept-encoding: deflate, gzip, br
> user-agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0
> accept: text/event-stream
> accept-language: en-US,en;q=0.5
> origin: https://domain2.localhost:4443
> connection: keep-alive
> referer: https://domain2.localhost:4443/
> sec-fetch-dest: empty
> sec-fetch-mode: cors
> sec-fetch-site: cross-site
> pragma: no-cache
> cache-control: no-cache
> te: trailers
> 
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* Connection state changed (MAX_CONCURRENT_STREAMS == 250)!
< HTTP/2 200 
< access-control-allow-credentials: true
< access-control-allow-origin: https://domain2.localhost:4443
< alt-svc: h3=":443"; ma=2592000,h3-34=":443"; ma=2592000,h3-32=":443"; ma=2592000,h3-29=":443"; ma=2592000
< cache-control: private, no-cache, no-store, must-revalidate, max-age=0
< content-security-policy: default-src 'self' mercure.rocks cdn.jsdelivr.net
< content-type: text/event-stream
< expire: 0
< pragma: no-cache
< server: Caddy
< x-accel-buffering: no
< x-content-type-options: nosniff
< x-frame-options: DENY
< x-xss-protection: 1; mode=block
< date: Mon, 30 Aug 2021 18:11:49 GMT
< 
:
```

**Comment by dunglas** [comment]
There is already a main issue: your tests domains are not sharing the same domain or subdomain. This will prevent the host to work.

Can you try with `mercure.votes.wolbodo` and `votes.wolbodo` to see if the issue persists?
Not using HTTPS can indeed also cause issues in some cases.

**Comment by neoel** [comment]
But that shouldn't make the cors fail right? I should be able to have domain-x.com be allowed by mercure on domainy.com

**Comment by dunglas** [comment]
I noticed the problem. There is an extra / in the allowed origin in your config.

`cors_origins http://votes.wolbodo/` should be `cors_origins http://votes.wolbodo`. And this fixes the issue according to my testing.

P.S: HTTP 2 works only (at least in browsers) if the connection is encrypted so it's normal, but it's not related here, the CORS library we use (https://github.com/gorilla/handlers) should work with both HTTP 1 and HTTP 2.

**Comment by TheoD02** [comment]
Hi I had the same problem,

Problem was i using 80 port instead of 8000

Here is my configuration you test with it

NGINX (port 80) :

```
  http:
    image: nginx:1-alpine
    environment:
      - VIRTUAL_HOST=gui.web.localhost
    networks:
        - reverse-proxy-net
        - default
    depends_on:
      - php
    volumes:
      - ./docker/nginx/vhost.conf:/etc/nginx/conf.d/default.conf:ro
      - ./app/public/:/srv/public/:ro
```

Caddy configuration
```
caddy:
    container_name: mercure
    image: dunglas/mercure
    restart: unless-stopped
    environment:
      # Uncomment the following line to disable HTTPS
      SERVER_NAME: 'mercure.gui.web.localhost'
      MERCURE_PUBLISHER_JWT_KEY: '!ChangeMe!'
      MERCURE_SUBSCRIBER_JWT_KEY: '!ChangeMe!'
    # Uncomment the following line to enable the development mode
    #command: /usr/bin/caddy run -config /etc/caddy/Caddyfile.dev
      MERCURE_EXTRA_DIRECTIVES: |-
        cors_origins "http://127.0.0.1" "http://mercure.gui.web.localhost"
        anonymous
    ports:
      - "8000:80"
      - "443:443"
    volumes:
      - caddy_data:/data
      - caddy_config:/config
```


JS Event
```
       // in event source need to have mercure.gui.web.localhost:8000 or 127.0.0.1:8000
        const eventSource = new EventSource('http://mercure.gui.web.localhost:8000/.well-known/mercure?topic=' + encodeURIComponent(`http://localhost/myTopic`));
```

PHP Send 
```
$update = new Update('http://localhost/myTopic', json_decode($response);
$hub->publish($update);
```

**Comment by Kern046** [comment]
Hi same problem here.

I made a [reproducer](https://github.com/Kern046/mercure-cors-reproducer) (my first one :grimacing: ).

I tried multiple values for the `CORS_ALLOWED_ORIGINS` environment variable as I saw it used in other issues and SO threads despite that I can't see it use in the Mercure Caddyfile :shrug: ).
I also tried the `MERCURE_EXTRA_DIRECTIVE` as @TheoD02 suggested but I fail to enable `CORS` in my project.

Here is my `docker-compose.yml` : 

```yaml
version: '3.6'

services:
    mercure:
        image: dunglas/mercure
        restart: unless-stopped
        environment:
            MERCURE_PUBLISHER_JWT_KEY: '!ChangeMe!'
            MERCURE_SUBSCRIBER_JWT_KEY: '!ChangeMe!'
            CORS_ALLOWED_ORIGINS: "https://local.mercure.wip"
            ALLOW_ANONYMOUS: 1
            SERVER_NAME: ":80"
            MERCURE_EXTRA_DIRECTIVES: |-
                cors_origins "https://local.mercure.wip"
                anonymous
        ports:
            - '80:80'
            - '443:443'
            - '20019:2019'
```

My JS code :

```js
const url = new URL('https://127.0.0.1/.well-known/mercure');
url.searchParams.append('topic', 'https://example.com/users/{username}');

const eventSource = new EventSource(url, {
    withCredentials: true
});

// The callback will be called every time an update is published
eventSource.onmessage = e => console.log(e); // do something with the payload
```

When calling `https://127.0.0.1/.well-known/mercure?topic=https://example.com/users/{username}` I receive no CORS Headers.

I'll be glad to dig in next week if this issue is relevant :).

**Comment by danielspk** [comment]
Same issue with docker.

Docker command:

```sh
docker run --rm --name=mercureprod -e MERCURE_PUBLISHER_JWT_KEY='!ChangeMe!' -e MERCURE_SUBSCRIBER_JWT_KEY='!ChangeMe!' -e MERCURE_EXTRA_DIRECTIVES='cors_origins http://localhost:3001' -p 80:80 -p 443:443 dunglas/mercure
```

Publish work!, but...

Javascript code subscriber in http://localhost:3001

```js
const url = new URL('https://localhost/.well-known/mercure');
url.searchParams.append('topic', 'https://demo.com/scan/{uuid}');

const eventSource = new EventSource(url, { withCredentials: true });

eventSource.onmessage = e => console.log(JSON.parse(e.data));
```

Error in console:

Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://localhost/.well-known/mercure?topic=https%3A%2F%2Fdemo.com%2Fscan%2F%7Buuid%7D. (Reason: CORS request did not succeed).

**Comment by dunglas** [comment]
@danielspk I just used the following commands in it works properly (no CORS error, the excepted 401):

```
$ docker run --rm --name=mercureprod -e MERCURE_PUBLISHER_JWT_KEY='!ChangeMe!' -e MERCURE_SUBSCRIBER_JWT_KEY='!ChangeMe!' -e MERCURE_EXTRA_DIRECTIVES='cors_origins http://localhost:3001' -p 80:80 -p 443:443 dunglas/mercure
$ python3 -m http.server 3001
```

The content of my `index.html` file:

```html
<script>
const url = new URL('https://localhost/.well-known/mercure');
url.searchParams.append('topic', 'https://demo.com/scan/{uuid}');

const eventSource = new EventSource(url, { withCredentials: true });

eventSource.onmessage = e => console.log(JSON.parse(e.data));
</script>
```

Be sure that:

1. you use the latest version (run `docker pull dunglas/mercure`
2. you accepted the self-signed certificate by visiting the URL of the hub directly in your browser


Generally speaking, it's sure that this feature works properly. I tested it (again) extensively today. With and without Docker (docs dedicated to Symfony coming) and it works as excepted. Also, and **all** previous reports I got and investigated were false positives and/or misconfigurations (last example in date: https://github.com/dunglas/mercure/issues/566).

Also, it's almost impossible to help you fix your problems without a proper reproducer (ideally a repository containing a Docker configuration). I cannot help at all without that.

**Comment by dunglas** [comment]
@Kern046 issue is (at least) because the hub is configured to use HTTP (`SERVER_NAME: :80`) but the EventSource subscribers to `https://127.0.0.1/.well-known/mercure` (extra s). Also, be careful if you use HTTPS, you **must** use the domain name `localhost` then, not the IP address. To use the IP address (not recommended), you must modify the `Caddyfile` to add extra unsafe options: https://caddy.community/t/https-on-a-local-ip-address-with-docker/11096/9

**Comment by acelaya** [comment]
I see many people reporting CORS-related issues here, but I don't see a clear solution other than some typos here and there, so sorry if this was actually solved already and I'm just repeating the same.

The thing is that for me it works with the legacy server, but not the caddy one.

If I use a docker container with docker-compose like this, it works:

```yaml
mercure:
    image: dunglas/mercure:legacy-v0.13
    ports:
        - "3080:80"
    environment:
        CORS_ALLOWED_ORIGINS: "https://app.shlink.io http://localhost:3000 http://127.0.0.1:3000"
        JWT_KEY: "mercure_jwt_key"
        USE_FORWARDED_HEADERS: "1"
```

But if I use the non-legacy image, like this, then I see CORS errors when trying to subscribe from the web app:

```yaml
mercure:
    image: dunglas/mercure:v0.13
    ports:
        - "3080:80"
    environment:
        MERCURE_PUBLISHER_JWT_KEY: mercure_jwt_key
        MERCURE_SUBSCRIBER_JWT_KEY: mercure_jwt_key
        MERCURE_EXTRA_DIRECTIVES: |
            cors_origins https://app.shlink.io http://localhost:3000 http://127.0.0.1:3000
```

`Access to fetch at 'http://localhost:3080/.well-known/mercure?topic=https%3A%2F%2Fshlink.io%2Fnew-orphan-visit' from origin 'https://app.shlink.io' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: Redirect is not allowed for a preflight request.`

Also, I haven't been able to find a clear migration guide in the docs, that explains the corresponding config from the legacy server to caddy directives, so maybe the problem is that I'm missing something.

**Comment by acelaya** [comment]
Ok, just found out what was missing for me. Since I'm accessing mercure without https, I needed to add this env var `SERVER_NAME: ":80"`

**Comment by acelaya** [comment]
I'm actually reading the `Redirect is not allowed for a preflight request.` part of the error, and I'm starting to think that mercure is trying to return a redirect from http to https for the preflight request (`OPTIONS`), causing many of the CORS errors posted here or in #442.

That would explain why it started to work for me once I disabled https.

**Comment by dunglas** [comment]
Good catch @acelaya! Indeed the default Caddy behavior (which is a good one IMHO) is to forward everything from HTTP to HTTPS if HTTPS is enabled. Maybe should we make this bold in our own docs.

**Comment by acelaya** [comment]
The problem is that, according to the spec, the preflight request should not result in a redirect. That's why the browser is throwing that error.

See https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSExternalRedirectNotAllowed

**Comment by dunglas** [comment]
@acelaya The error is intended. It can only occur when there is a misconfiguration. If HTTPS is enabled, then HTTPS must be used (actually, according to the Mercure spec, HTTPS must always be used).

**Comment by acelaya** [comment]
Got it.

Then yeah, documenting it somewhere should be enough.

**Comment by halfred** [comment]
Hi I'm also having a CORS problem with the latest docker caddy version of mercure

It's working as expected without https in local dev environnment but not in prod

I'm using a react app as a client and both the app and mercure are using traefik 2

Here is my docker-compose file

```yaml
  # MERCURE #
  mercure:
    build:
      context: ./docker/mercure
      #in Dockerfile : FROM dunglas/mercure
    volumes:
      - .data/:/data
      - .config/:/config
    container_name: mercure
    environment:
      MERCURE_PUBLISHER_JWT_KEY: 'my-key'
      MERCURE_SUBSCRIBER_JWT_KEY: 'my-key'
    labels:
      - "traefik.enable=true"

      - "traefik.http.routers.mercure.rule=Host(`mercure.domain.com`)"
      - "traefik.http.routers.mercure.entrypoints=websecure"
      - "traefik.http.routers.mercure.tls.certresolver=myresolver"

  # REACT #
  react:
    build:
      context: ./docker/app
    container_name: react
    env_file:
      - env_file.prod
    labels:
      - "traefik.enable=true"
     # all 80 requests are redirected to 443
      - "traefik.http.routers.http_node.rule=Host(`react.domain.com`)"
      - "traefik.http.routers.http_node.entrypoints=web"
     
      - "traefik.http.routers.hhm-node.rule=Host(`react.domain.com`)"
      - "traefik.http.routers.hhm-node.entrypoints=websecure"
      - "traefik.http.routers.hhm-node.tls.certresolver=myresolver"
```

and the Caddyfile

```
{$SERVER_NAME:localhost}

log

route {
	encode zstd gzip

	mercure {
		# Transport to use (default to Bolt)
		transport_url {$MERCURE_TRANSPORT_URL:bolt://mercure.db}
		# Publisher JWT key
		publisher_jwt {env.MERCURE_PUBLISHER_JWT_KEY} {env.MERCURE_PUBLISHER_JWT_ALG}
		# Subscriber JWT key
		subscriber_jwt {env.MERCURE_SUBSCRIBER_JWT_KEY} {env.MERCURE_SUBSCRIBER_JWT_ALG}
		cors_origins https://react.domain.com
		publish_origins *
		ui
		subscriptions
		# Extra directives
		{$MERCURE_EXTRA_DIRECTIVES}
	}

	respond /healthz 200

	respond "Not Found" 404
}

```


I tried a lot of changes but nothing worked.
I even tried with `cors_origins *` in the Caddyfile but the CORS headers where still missing and firefox complained in my app console

Am I missing something or is it a bug?

Thanks

**Comment by halfred** [comment]
Maybe it's a Caddy issue

I noticed that a curl in dev with https (ignoring the local cert check) is opening a connection as expected.
But in prod, curl returns an http code 308.

Postman displays the error as: 
`Error: Exceeded maxRedirects. Probably stuck in a redirect loop https://mercure.domain.com/.well-known/mercure`

**Comment by halfred** [comment]
Can someone please help me ?
I haven't progressed a bit since 2 weeks :(

**Comment by dunglas** [comment]
Sorry I don't use Caddy... Could you provide a reproducer in a repository so I can try to hack it?

**Comment by thund3rb1rd78** [comment]
> Can someone please help me ? I haven't progressed a bit since 2 weeks :(

halfred, I don't know if it will help, as I use different technologies (it seems) than you. But, perhaps you can be inspired by something you see. This is a link to my skeleton I use that works in production and local encrypted with auto-ssl for symfony/mercure/caddy/mariadb: https://github.com/thund3rb1rd78/symfony-mercure-website-skeleton-dockerized
good luck

**Comment by halfred** [comment]
Thanks @dunglas and @thund3rb1rd78 for your suggestions
I'm back on the project after a few days.
I don't understand why but now it's not working anymore even in dev where I thought that I disabled the cors checks

I checked the skeleton, based on dunglas' one and tried to identify what's wrong in my conf but I couldn't find anything relevant.
I'll continue my investigations in the next days and I'll post the results if I find something useful for the community.

In the mean time, here is my dev Caddyfile. Maybe someone will find an obvious mistake

```
{
    # Debug
    {$DEBUG}
    # HTTP/3 support
    servers {
        protocol {
            experimental_http3
        }
    }
}

{$SERVER_NAME:localhost}

log

route {
	redir / /.well-known/mercure/ui/
	encode zstd gzip

	mercure {
		# Transport to use (default to Bolt)
		transport_url {$MERCURE_TRANSPORT_URL:bolt://mercure.db}
		# Publisher JWT key
		publisher_jwt {env.MERCURE_PUBLISHER_JWT_KEY} {env.MERCURE_PUBLISHER_JWT_ALG}
		# Subscriber JWT key
		subscriber_jwt {env.MERCURE_SUBSCRIBER_JWT_KEY} {env.MERCURE_SUBSCRIBER_JWT_ALG}
		# Permissive configuration for the development environment
		cors_origins *
		publish_origins *
		ui
		anonymous
		subscriptions
		# Extra directives
		{$MERCURE_EXTRA_DIRECTIVES}
	}

	respond /healthz 200
	respond "Not Found" 404
}

```

**Comment by halfred** [comment]
So I've finally manage to make it "work" in dev env.
But it requires a trick by enabling the "CORS Everywhere" extension on firefox (which is obviously cheating)

I think that I properly followed the official mercury documentation about cors : https://mercure.rocks/docs/hub/troubleshooting#cors-issues
From my previous message, I changed the Caddyfile line according to the doc and to my env.
In dev, on the address http://localhost:8082 runs my js app and on http://localhost:8083 runs the mercury hub
So I set:

```
cors_origins http://localhost:8083
```
But as I said, without the extension enabled, the request is blocked by firefox.

I'll try to post on stack overflow to get some help, in case it's not a bug.

**Comment by halfred** [comment]
I've found 2 mistakes and a workaround and now everything seems to work as expected.

The mistake was the port in the cors url. In my case it must be 8082 (the port of my js app) instead of 8083 (the mercure hub port).

And the workaround for the cors header in caddy:
I gave up the cors_origin directive in the mercure block of the Caddyfile.
I used instead a caddy header setter.

Here is my dev Caddyfile:

```
{
	{$GLOBAL_OPTIONS}
	debug
}

{$SERVER_NAME:localhost}

log

route {
	#redir / /.well-known/mercure/ui/
	encode zstd gzip

	@origin header Origin http://localhost:8082
        header @origin Access-Control-Allow-Origin "http://localhost:8082"
        header @origin Access-Control-Request-Method GET
        header Access-Control-Allow-Credentials "true"

	mercure {
		# Transport to use (default to Bolt)
		transport_url {$MERCURE_TRANSPORT_URL:bolt://mercure.db}
		# Publisher JWT key
		publisher_jwt {env.MERCURE_PUBLISHER_JWT_KEY} {env.MERCURE_PUBLISHER_JWT_ALG}
		# Subscriber JWT key
		subscriber_jwt {env.MERCURE_SUBSCRIBER_JWT_KEY} {env.MERCURE_SUBSCRIBER_JWT_ALG}
		# Permissive configuration for the development environment
		publish_origins *
		ui
		anonymous
		#subscriptions
		# Extra directives
		{$MERCURE_EXTRA_DIRECTIVES}
	}

	respond /healthz 200
	respond "Not Found" 404
}
```

Finally, my initial http code 308 redirection loop  was due to the port 80 disabled.
As traefik handles the https security, I can leave the unencrypted 80 port opened in prod as in dev:
```
#in the docker-compose file
environment:
     SERVER_NAME: ':80'
```
